common/lcs/src/de.rs

```diff
@@ -48,21 +48,25 @@ pub fn from_bytes_seed<'a, T>(seed: T, bytes: &'a [u8]) -> Result<T::Value>
 where
     T: DeserializeSeed<'a>,
 {
-    let mut deserializer = Deserializer::new(bytes);
+    let mut deserializer = Deserializer::new(bytes, crate::MAX_OBJECT_DEPTH);
     let t = seed.deserialize(&mut deserializer)?;
     deserializer.end().map(move |_| t)
 }
 
 /// Deserialization implementation for LCS
 struct Deserializer<'de> {
     input: &'de [u8],
+    max_remaining_depth: usize,
 }
 
 impl<'de> Deserializer<'de> {
     /// Creates a new `Deserializer` which will be deserializing the provided
     /// input.
-    fn new(input: &'de [u8]) -> Self {
-        Deserializer { input }
+    fn new(input: &'de [u8], max_remaining_depth: usize) -> Self {
```

Comment:
I suspect you want to check that max_remaining_depth > 0.
Since the last argument will be difficult to parametrize for newcomers (or people who don't read docs) *and* this is safety-sensitive, would it make sense to have it be an `Into<Option<usize>>`? That way uncertain caller could pass `None`? Alternatively, would it be wise to implement `Default`?
I really want to avoid people putting a Very Large Number there because they don't know what a good value looks like. And I'm nearly rempted to recommend comparing `max_remaining_depth` with `MAX_OBJECT_DEPTH`, and issuing a warning when the former is bigger.
Same remark for `Serializer` ofc.

Reply:
@huitseeker These constructors are not `pub`.

Reply:
Do we expect every validator that would ever run LCS serialization to have/need the same stack size limits ? 

I suspect not.

If not, do we plan to get all validators on the same recursion depth limit ? If so, would that be the lowest common denominator among those validators, and would we be ok with the implication on max transaction/block/what-have-you size over a long period of time? How would we handle an upgrade?

I suspect we'll rather allow distinct validators to run with different values for this parameter (under controlled conditions). What I'm getting at is this constructor will probably not remain private forever.

Reply:
In this PR, the depth limit is part of the definition of LCS, so it is the same everywhere.

Reply:
I don't see how to do otherwise by the way, since we want to deterministic execution etc.

Reply:
Although in practice, it's probably fine to make this value larger in rare occasions.